File: kestrel-java/src/main/java/io/github/matthewjones372/kestrel/java/Results.java
package io.github.matthewjones372.kestrel.java;

import io.github.matthewjones372.kestrel.Clock;
import io.github.matthewjones372.kestrel.RunResult;
import io.github.matthewjones372.kestrel.StepName;
import io.github.matthewjones372.kestrel.Verdict;
import java.time.Duration;
import java.util.List;

/**
 * What a run measured, off core's own {@code RunResult} rather than a copy of
 * it. Every duration leaves here as a {@code java.time.Duration}, since
 * {@code kotlin.time.Duration} reaches Java as a bare {@code long} whose unit
 * the type no longer states.
 *
 * The clock defaults to {@link Clock#ResponseTime}, as it does in Kotlin: a
 * goal written against service time can be met by a generator that never sent
 * the load.
 */
public final class Results {

    private Results() {
    }

    /** Every goal the run declared, with what it measured and the margin it missed by. */
    public static List<Verdict> verdicts(RunResult result) {
        return Judgements.verdicts(result);
    }

    public static boolean ran(RunResult result, StepName step) {
        return result.ran(step.getName());
    }

    public static long count(RunResult result, StepName step) {
        return Reads.stats(result, step.getName()).getCount();
    }

    public static long ok(RunResult result, StepName step) {
        return Reads.stats(result, step.getName()).getOk().getCount();
    }

    public static long failed(RunResult result, StepName step) {
        return Reads.stats(result, step.getName()).getFailed().getCount();
    }

    public static Duration p50(RunResult result, StepName step) {
        return p50(result, step, Clock.ResponseTime);
    }

    public static Duration p50(RunResult result, StepName step, Clock clock) {
        return Reads.percentile(result, step.getName(), 50.0, clock);
    }

    public static Duration p95(RunResult result, StepName step) {
        return p95(result, step, Clock.ResponseTime);
    }

    public static Duration p95(RunResult result, StepName step, Clock clock) {
        return Reads.percentile(result, step.getName(), 95.0, clock);
    }

    public static Duration p99(RunResult result, StepName step) {
        return p99(result, step, Clock.ResponseTime);
    }

    public static Duration p99(RunResult result, StepName step, Clock clock) {
        return Reads.percentile(result, step.getName(), 99.0, clock);
    }

    public static Duration max(RunResult result, StepName step) {
        return max(result, step, Clock.ResponseTime);
    }

    public static Duration max(RunResult result, StepName step, Clock clock) {
        return Reads.max(result, step.getName(), clock);
    }
}

File: proofload-scala/src/main/scala/io/github/matthewjones372/proofload/scala/Results.scala
package io.github.matthewjones372.proofload.scala

import io.github.matthewjones372.proofload.Clock
import io.github.matthewjones372.proofload.LimitsKt
import io.github.matthewjones372.proofload.RunResult
import io.github.matthewjones372.proofload.RunResultKt
import io.github.matthewjones372.proofload.StepName
import io.github.matthewjones372.proofload.java.Results
import _root_.scala.concurrent.duration.FiniteDuration

/**
 * What one step measured, on one clock, read off core's own `RunResult`
 * through the Java facade. It holds no number of its own and computes none.
 *
 * The clock starts at response time, as it does in Kotlin: a goal written
 * against service time can be met by a generator that never sent the load.
 */
final class Measured private[scala] (result: RunResult, step: StepName, clock: Clock):

  def responseTime: Measured = Measured(result, step, Clock.ResponseTime)

  def serviceTime: Measured = Measured(result, step, Clock.ServiceTime)

  def ran: Boolean = Results.ran(result, step)

  def count: Long = Results.count(result, step)

  def ok: Long = Results.ok(result, step)

  def failed: Long = Results.failed(result, step)

  def p50: FiniteDuration = asScala(Results.p50(result, step, clock))

  def p95: FiniteDuration = asScala(Results.p95(result, step, clock))

  def p99: FiniteDuration = asScala(Results.p99(result, step, clock))

  def max: FiniteDuration = asScala(Results.max(result, step, clock))

extension (result: RunResult)

  def apply(step: StepName): Measured = Measured(result, step, Clock.ResponseTime)

  /** Whether the generator's own lateness is a material part of the tail this run reports. */
  def fellBehind: Boolean = RunResultKt.fellBehind(result)

  /** Whether it lost the schedule outright, which makes the rate on the page one nobody offered. */
  def lostGround: Boolean = RunResultKt.lostGround(result)

  /** Whether this process ran out of its own room, in which case the run measured the injector. */
  def ranOutOfRoom: Boolean = LimitsKt.ranOutOfRoom(result)
